File: vol1/ch11/LeetCode238-2.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n=nums.size();
        vector<int> ans(n);
        ans[0]=nums[0];                //求前缀积
        for (int i=1;i<n;i++)
            ans[i]=ans[i-1]*nums[i];
        ans[n-1]=ans[n-2];
        int postd=nums[n-1];    //求后缀积
        for (int k=n-2;k>=1;k--) {
            ans[k]=ans[k-1]*postd;
            postd*=nums[k];
        }
        ans[0]=postd;
        return ans;
    }
};

File: vol1/ch11/LeetCode238-2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n=len(nums)
        ans=[0]*n
        ans[0]=nums[0]                #求前缀积
        for i in range(1,n):
            ans[i]=ans[i-1]*nums[i]
        ans[n-1]=ans[n-2]
        postd=nums[n-1]                 #求后缀积
        for k in range(n-2,0,-1):
            ans[k]=ans[k-1]*postd
            postd*=nums[k]
        ans[0]=postd
        return ans
